File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int N, M, max = Integer.MIN_VALUE;
    static int[][] board;
    public static void main(String[] args) throws IOException {
        setVariables();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (j + 3 < M) {
                    max = Math.max(max, board[i][j] + board[i][j + 1] + board[i][j + 2] + board[i][j + 3]);
                }
                if (i + 3 < N) {
                    max = Math.max(max, board[i][j] + board[i + 1][j] + board[i + 2][j] + board[i + 3][j]);
                }
                if (i + 1 < N && j + 1 < M) {
                    max = Math.max(max, board[i][j] + board[i + 1][j] + board[i][j + 1] + board[i + 1][j + 1]);
                }
                if (i + 1 < N && j + 2 < M) {
                    max = Math.max(max, twoByThree(i, j));
                }
                if (i + 2 < N && j + 1 < M) {
                    max = Math.max(max, threeByTwo(i, j));
                }
            }
        }
        System.out.println(max);
    }
    static void setVariables() throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        board = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }
    static int twoByThree(int i, int j) {
        int answer = Integer.MIN_VALUE;
        int box = 0;
        for (int k = 0; k < 2; k++) {
            for (int l = 0; l < 3; l++) {
                box += board[i + k][j + l];
            }
        }
        int[][] rowIndices = {{0, 0}, {0, 0}, {0, 0}, {0, 1}, {1, 0}, {1, 1}, {1, 1}, {1, 1}},
                colIndices = {{0, 1}, {0, 2}, {1, 2}, {0, 2}, {0, 2}, {0, 1}, {0, 2}, {1, 2}};
        for (int k = 0; k < 8; k++) {
            answer = Math.max(answer, box - board[i + rowIndices[k][0]][j + colIndices[k][0]]
                                - board[i + rowIndices[k][1]][j + colIndices[k][1]]);
        }
        return answer;
    }
    static int threeByTwo(int i, int j) {
        int answer = Integer.MIN_VALUE;
        int box = 0;
        for (int k = 0; k < 3; k++) {
            for (int l = 0; l < 2; l++) {
                box += board[i + k][j + l];
            }
        }
        int[][] rowIndices = {{0, 1}, {0, 2}, {1, 2}, {0, 2}, {0, 2}, {0, 1}, {0, 2}, {1, 2}},
                colIndices = {{0, 0}, {0, 0}, {0, 0}, {0, 1}, {1, 0}, {1, 1}, {1, 1}, {1, 1}};
        for (int k = 0; k < 8; k++) {
            answer = Math.max(answer, box - board[i + rowIndices[k][0]][j + colIndices[k][0]]
                    - board[i + rowIndices[k][1]][j + colIndices[k][1]]);
        }
        return answer;
    }
}

File: 백준/Gold/14500. 테트로미노/테트로미노.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

#define FAST_IO cin.tie(NULL)->ios::sync_with_stdio(false);

using namespace std;

#define REP(i, a, b) for (int i = a; i <= b; i++)

int n, m;
vector<vector<int>> board;

int two_by_three(int i, int j) {
    int answer = INT_MIN, box = 0;
    REP(k, 0, 1) REP(l, 0, 2) box += board[i + k][j + l];

    int row_indices[8][2] = {
            {0, 0}, {0, 0}, {0, 0}, {0, 1},
            {1, 0}, {1, 1}, {1, 1}, {1, 1}
    }, col_indices[8][2] = {
            {0, 1}, {0, 2}, {1, 2}, {0, 2},
            {0, 2}, {0, 1}, {0, 2}, {1, 2}
    };

    REP(k, 0, 7) {
        int r1 = i + row_indices[k][0], c1 = j + col_indices[k][0],
            r2 = i + row_indices[k][1], c2 = j + col_indices[k][1];

        int current = box - (board[r1][c1] + board[r2][c2]);
        answer = max(answer, current);
    }

    return answer;
}

int three_by_two(int i, int j) {
    int answer = INT_MIN, box = 0;
    REP(k, 0, 2) REP(l, 0, 1) box += board[i + k][j + l];

    int row_indices[8][2] = {
            {0, 1}, {0, 2}, {1, 2}, {0, 2},
            {0, 2}, {0, 1}, {0, 2}, {1, 2}
    }, col_indices[8][2] = {
            {0, 0}, {0, 0}, {0, 0}, {0, 1},
            {1, 0}, {1, 1}, {1, 1}, {1, 1}
    };

    REP(k, 0, 7) {
        int r1 = i + row_indices[k][0],
            c1 = j + col_indices[k][0],
            r2 = i + row_indices[k][1],
            c2 = j + col_indices[k][1];

        int current = box - (board[r1][c1] + board[r2][c2]);
        answer = max(answer, current);
    }
    return answer;
}

int main() {
    FAST_IO

    cin >> n >> m;
    board.resize(n, vector<int>(m));
    REP(i, 0, n - 1) REP(j, 0, m - 1) cin >> board[i][j];

    int max_val = INT_MIN;
    REP(i, 0, n - 1) {
        REP(j, 0, m - 1) {
            // 가로로 4칸
            if (j + 3 < m) {
                int sum4 = board[i][j] + board[i][j+1] + board[i][j+2] + board[i][j+3];
                max_val = max(max_val, sum4);
            }
            // 세로로 4칸
            if(i + 3 < n){
                int sum4 = board[i][j] + board[i+1][j] + board[i+2][j] + board[i+3][j];
                max_val = max(max_val, sum4);
            }
            // 2×2 정사각형
            if(i + 1 < n && j + 1 < m){
                int sumSquare = board[i][j] + board[i+1][j] + board[i][j+1] + board[i+1][j+1];
                max_val = max(max_val, sumSquare);
            }
            // 2×3 형태 (테트리스 모양)
            if(i + 1 < n && j + 2 < m){
                max_val = max(max_val, two_by_three(i, j));
            }
            // 3×2 형태 (테트리스 모양)
            if(i + 2 < n && j + 1 < m){
                max_val = max(max_val, three_by_two(i, j));
            }
        }
    }

    cout << max_val;

    return EXIT_SUCCESS;
}